### PR TITLE
Blood moon rising

### PR DESCRIPTION
**Pull request by MadmanMartian**
**Adds the blood moon event for admins to trigger via the secrets tab**
- Ruins all the APCs
- Makes space red
- Litters the station with gravestones
- - The gravestones spawn horrors depending on their location
- - Vampires are the worst to encounter in the dark
- - The kitchen can hold its own against the skeletal horde
- - Stay away from the supermatter shade
- - The chapel is safe from the initial tide, but will not hold for long.

Still testing and adding more mobs as their ideas are formed/their sprites are made.

:cl:
* rscadd: Adds the blood moon world event. Board up the windows and hope you have what it takes to survive.

### CODE REVIEW

**Comment by Shadowmech88** [comment]
What about a captain mob that, as it wanders, travels between departments by opening hand teleporter portals?
Or even better, the mob stays put, but can attack anyone anywhere on the station by opening a portal and shooting the captain's laser at them through it.

**Comment by Shadowmech88** [comment]
Oh also, I think the button should make ghosts visible like they are during Nar-Sie.

**Comment by ultraplen** [comment]
What about grues? Sounds like a decent place to use em.

**Comment by Shadowmech88** [comment]
Grues are player-controlled-only, I think this event is strictly NPC enemies.

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm [hunk 1]
```diff
@@ -0,0 +1,395 @@
+/*
+	Halloween related mobs
+*/
+
+/mob/living/simple_animal/hostile/humanoid/kitchen
+	faction = "kitchen_nightmares"
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/human/halloween.dm:6]: starring gordon ramsey

**Comment by SonixApache** [comment]
post sprites and such

File: code/game/gamemodes/endgame/halloween/gravestone.dm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	gravestone_flick()
+	..()
+
+/obj/effect/gravestone/proc/gravestone_flick()
+	return
+
+/obj/effect/gravestone/stone
+	icon_state = "gravestone_stone"
+
+var/global/list/halloween_spawns = list(
```
Comment (PJB3005) [inline code/game/gamemodes/endgame/halloween/gravestone.dm]: `/global` does nothing here.

File: code/game/gamemodes/endgame/halloween/gravestone.dm [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+/obj/effect/gravestone/halloween/gravestone_flick()
+	flick("[icon_state]_new", src)
+
+/obj/effect/gravestone/halloween/proc/spawn_enemies()
+	var/our_area = get_area_string()
+
+	if(our_area == "space" || our_area == "chapel")
+		animate(src, alpha = 0, time = 1 SECONDS)
+		spawn(1 SECONDS)
```
Comment (PJB3005) [inline code/game/gamemodes/endgame/halloween/gravestone.dm]: The code after the deletion will still run, though the things will be spawned in nullspace.

Comment (PJB3005) [reply code/game/gamemodes/endgame/halloween/gravestone.dm]: Also, give the proc `set waitfor = FALSE` and change the `spawn()` calls into `sleep()`

File: code/game/gamemodes/endgame/halloween/gravestone.dm [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+			qdel(src)
+
+	var/list/possible_spawns = halloween_spawns["default"] + halloween_spawns[our_area]
+
+	spawn(rand(15 SECONDS, 30 SECONDS))
+		var/to_spawn = pick(possible_spawns)
+		if(to_spawn)
```
Comment (PJB3005) [inline code/game/gamemodes/endgame/halloween/gravestone.dm:63]: Could `to_spawn` ever return null here?

File: code/game/gamemodes/endgame/halloween/gravestone.dm [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			new to_spawn(get_turf(src))
+		animate(src, alpha = 0, time = 3 SECONDS)
+		spawn(3 SECONDS)
+			qdel(src)
+
+/obj/effect/gravestone/halloween/proc/get_area_string()
+	var/our_area_type = "default"
```
Comment (PJB3005) [inline code/game/gamemodes/endgame/halloween/gravestone.dm]: Make this use `.` instead. It's the default return value of the proc. Just set it directly and remove the return.

**Comment by Blithering** [comment]
superman
Supermatter needs hugs too

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+	melee_damage_lower = 20
+	melee_damage_upper = 35
+	attacktext = "smashes their cheese-covered gauntlets into"
+	attack_sound = 'sound/weapons/heavysmash.ogg'
+	corpse = null
+
+/mob/living/simple_animal/hostile/humanoid/kitchen/poutine/bullet_act(var/obj/item/projectile/P)
+	if(istype(P, /obj/item/projectile/beam))
+		P.damage /= rand(1.2,3)
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/hostile/human/halloween.dm]: `rand` doesn't support non-natural numbers:

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+		if(prob(45))
+			visible_message("<span class = 'warning'>\The [P] has a reduced effect on \the [src]!</span>")
+
+	return (..(P))
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/hostile/human/halloween.dm]: Why wrap it in parentheses?

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+
+/mob/living/simple_animal/hostile/humanoid/kitchen/poutine/Die()
+	for(var/i=0;i<3;i++)
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/hostile/human/halloween.dm]: Use `for (var/i = 1 to 3)` instead.

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+		var/to_spawn = pick(/obj/item/weapon/reagent_containers/food/snacks/poutine, /obj/item/weapon/reagent_containers/food/snacks/poutinedangerous,\
+		/obj/item/weapon/reagent_containers/food/snacks/poutinebarrel)
```
Comment (PJB3005) [inline code/modules/mob/living/simple_animal/hostile/human/halloween.dm]: Indent this line so it's on the same level as the other types.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/hostile/human/halloween.dm (hunk 6 above)
Comment (Shadowmech88) [reply code/modules/mob/living/simple_animal/hostile/human/halloween.dm]: You could do `rand(12,30)/10`.

**Comment by Exxion** [comment]
What moon?

File: code/game/gamemodes/endgame/halloween/gravestone.dm (hunk 2 above)
Comment (MadmanMartian) [reply code/game/gamemodes/endgame/halloween/gravestone.dm]: so, can it just be accessed by things in this file? If so that's pleasant

File: code/game/gamemodes/endgame/halloween/gravestone.dm (hunk 4 above)
Comment (MadmanMartian) [reply code/game/gamemodes/endgame/halloween/gravestone.dm:63]: It did, and I was unable to reproduce why and/or how. Happened both on the test_tiny and boxstation

Comment (PJB3005) [reply code/game/gamemodes/endgame/halloween/gravestone.dm:63]: Huh. Weird.

File: code/game/gamemodes/endgame/halloween/gravestone.dm (hunk 2 above)
Comment (PJB3005) [reply code/game/gamemodes/endgame/halloween/gravestone.dm]: Nah, `/global` literally just does nothing when used for global variables.

`/global` (and its alias, `/static`) make a static variable out of object or proc variables.

Comment (DamianX) [reply code/game/gamemodes/endgame/halloween/gravestone.dm]: It's the opposite, it can be accessed from everywhere since it's not defined under any datum

**Comment by N3X15** [comment]
Conflicts

**Comment by Exxion** [comment]
No, seriously
What moon?

**Comment by MadmanMartian** [comment]
@Exxion we're working on it

May not be in in time for halloween, FTL13 code is pretty cryptic, and so is clusterfack when explaining how parallax works

**Comment by Exxion** [comment]
So ask someone else lol

There are a bunch of images on each person's screen moved around as they move, on the third- and fourth-lowest planes. Space is rendered as white squares on the second-lowest plane. The images have `BLEND_MULTIPLY`, so they appear unaltered where there's white (space) and don't appear where there's black (blocked turfs) (because each of their color components is multiplied by the same one in the thing they overlay, divided by 255)

TL;DR if you want to change the color of parallax, just look at Cultcode
Which is presumably "give each person a `PLANE_MASTER` at `PLANE_SPACE` with `color` red

Also I lied about n<sup>th</sup>-lowest planes because actually any integer is a valid plane, but the DM ref says only -100 through 100 are valid and we just always sort of assumed that was true

**Comment by MadmanMartian** [comment]
Parallax already changes by copypasting the hells rising code.

I want to add a moon to it, like you're suggesting.

**Comment by Exxion** [comment]
Oh
One that stays still?

**Comment by MadmanMartian** [comment]
Kinda. I assume you'd add it to the bottom-most layer of some form.

I'm looking into it right now, as it seems FTL parallax and /vg/station parallax are two different beasts

**Comment by Exxion** [comment]
If you want the moon to move around with the bottom layer of parallax, yes, you'd add it to one of the bottom-layer parallax images and update everyone's parallax images. But it would likely look weird if you did that, as it would loop around your screen as it moved and stars would render in front of it.
I'd just make it stand still by just giving everyone a static moon image on the topmost parallax layer except dust.

File: code/game/gamemodes/endgame/halloween/halloween.dm [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+/datum/universal_state/halloween/proc/MiscSet()
+	for (var/obj/machinery/firealarm/alm in machines)
+		if (!(alm.stat & BROKEN))
+			alm.ex_act(2)
+		CHECK_TICK
+
+/datum/universal_state/halloween/proc/APCSet()
+	for (var/obj/machinery/power/apc/APC in power_machines)
+		if (!(APC.stat & BROKEN) && !istype(APC.areaMaster,/area/turret_protected/ai))
```
Comment (Kurfursten) [inline code/game/gamemodes/endgame/halloween/halloween.dm]: You need to exclude engineering like the station blackout event does unless you want this to singuloose

**Review by PJB3005 — APPROVED** [review]
